imports/server/mediasoup-api.ts

```diff
@@ -137,18 +137,43 @@ Meteor.publish('mediasoup:join', function (hunt, call, tab) {
     // sure that the client no longer trusts that the previous `call.join` is
     // active, and we can just take this as a signal to hasten that cleanup.
     // If the room is not yet created, create it.
-    await Peers.removeAsync({ hunt, call, tab });
+    const maybeOldPeer = await Peers.findOneAsync({ hunt, call, tab });
+    if (maybeOldPeer) {
+      Ansible.log(`Removing peer with same hunt/call/tab`, { peer: maybeOldPeer._id })
+      await Peers.removeAsync({ _id: maybeOldPeer._id, hunt, call, tab });
+    }
+
+    const peerCount = await Peers.find({ hunt, call }).countAsync();
+    // If we are a new joiner and would be the 8th (or more) peer, join the
+    // call starting out muted, because large calls can get noisy.
+    // In local development, make this just 3 because it's annoying to open that many browser tabs.
+    const crowdSize = Meteor.isDevelopment ? 3 : 8;
+    let initialPeerState: 'active' | 'muted' | 'deafened' = peerCount + 1 >= crowdSize ? 'muted' : 'active';
+    // But if we were previously in call, just restore whatever the previous
+    // mute state was.
+    if (maybeOldPeer) {
+      let oldPeerState;
+      if (maybeOldPeer.deafened) {
+        oldPeerState = 'deafened' as const;
+      } else if (maybeOldPeer.muted) {
+        oldPeerState = 'muted' as const;
+      } else {
+        oldPeerState = 'active' as const;
+      }
```

Comment:
OOC, does TypeScript back-propagate the types from these assignments to `oldPeerState`, or does it effectively end up as `any`?

Reply:
Yeah looks like it ends up as `any`: https://tsplay.dev/mL5qVW

Reply:
Ah no, my mistake - it is `any` when it's first declared, but the type is narrowed by having an exhaustive set of if statements, so when the actual subsequent assignment happens, it has a real type: https://tsplay.dev/w1APXw

Reply:
It appears to know the narrower type after the assignments in the flow.  (The type on line 155 is `any`, but by the time we reach line 163 it's `'active' | 'muted' | 'deafened'`.)